File: core/governance/gate_policy.py
"""
Constitutional Gate Policy — Python mirror of bizra-core/src/gate_policy.rs.

Unified enforcement for all constitutional threshold checks (Ihsan, SNR, ADL).

The same threshold check (e.g., ``ihsan < 0.95``) previously triggered 5 different
behaviors across modules. This module provides ONE canonical decision function that
all enforcement surfaces should call.

Standing on Giants: Al-Ghazali (Ihsan as obligation) · Deming (quality at source)
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from enum import Enum


class GatePolicy(Enum):
    """What happens when a constitutional threshold is violated.

    Ordered from most permissive to most restrictive:
    OBSERVE → FLAG → THROTTLE → REJECT
    """

    OBSERVE = "observe"
    FLAG = "flag"
    THROTTLE = "throttle"
    REJECT = "reject"


class GateAction(Enum):
    """The action taken after policy evaluation."""

    ALLOW = "allow"
    ALLOW_WITH_WARNING = "allow_with_warning"
    FLAGGED = "flagged"
    THROTTLED = "throttled"
    REJECTED = "rejected"


@dataclass(frozen=True)
class GateVerdict:
    """The result of applying a gate policy to a score."""

    score: float
    threshold: float
    passed: bool
    policy: GatePolicy
    action: GateAction


def env_gate_policy() -> GatePolicy:
    """Resolve gate policy from ``BIZRA_ENV`` environment variable.

    - ``BIZRA_ENV=prod`` or ``BIZRA_ENV=production`` → REJECT
    - All other values or unset → OBSERVE
    """
    env = os.environ.get("BIZRA_ENV", "").lower()
    if env in ("prod", "production"):
        return GatePolicy.REJECT
    return GatePolicy.OBSERVE


def apply_gate(
    score: float,
    threshold: float,
    policy: GatePolicy | None = None,
) -> GateVerdict:
    """Apply a gate policy to a score/threshold pair.

    This is the ONE function all enforcement surfaces should call.

    If ``policy`` is None, resolves from ``BIZRA_ENV``.

    Args:
        score: The measured quality score.
        threshold: The constitutional floor.
        policy: Enforcement policy. Defaults to env-resolved.

    Returns:
        GateVerdict with the decision.
    """
    if policy is None:
        policy = env_gate_policy()

    passed = score >= threshold
    if passed:
        action = GateAction.ALLOW
    elif policy == GatePolicy.OBSERVE:
        action = GateAction.ALLOW_WITH_WARNING
    elif policy == GatePolicy.FLAG:
        action = GateAction.FLAGGED
    elif policy == GatePolicy.THROTTLE:
        action = GateAction.THROTTLED
    else:
        action = GateAction.REJECTED

    return GateVerdict(
        score=score,
        threshold=threshold,
        passed=passed,
        policy=policy,
        action=action,
    )


# ── Wire 5: Gate Maturation ──────────────────────────────────────────────────


@dataclass
class MaturationThresholds:
    """Cycle counts at which the policy auto-promotes.

    Standing on Giants: Deming (PDCA maturation) · Lamport (safety liveness)
    """

    observe_to_flag: int = 100
    flag_to_throttle: int = 500
    throttle_to_reject: int = 1000


class GateMaturationPolicy:
    """Auto-promoting gate policy that hardens with accumulated evidence.

    Starts at OBSERVE and promotes through the GatePolicy ladder:
      OBSERVE → FLAG → THROTTLE → REJECT

    Each tick() increments the cycle counter. Promotion is monotonic —
    a gate never softens once hardened.
    """

    def __init__(self, thresholds: MaturationThresholds | None = None) -> None:
        self._thresholds = thresholds or MaturationThresholds()
        self._cycle_count = 0
        self._current = GatePolicy.OBSERVE

    def tick(self) -> GatePolicy:
        """Record one cycle and auto-promote if a threshold is crossed."""
        self._cycle_count += 1
        if (
            self._current == GatePolicy.OBSERVE
            and self._cycle_count >= self._thresholds.observe_to_flag
        ):
            self._current = GatePolicy.FLAG
        elif (
            self._current == GatePolicy.FLAG
            and self._cycle_count >= self._thresholds.flag_to_throttle
        ):
            self._current = GatePolicy.THROTTLE
        elif (
            self._current == GatePolicy.THROTTLE
            and self._cycle_count >= self._thresholds.throttle_to_reject
        ):
            self._current = GatePolicy.REJECT
        return self._current

    @property
    def current(self) -> GatePolicy:
        """Current active policy."""
        return self._current

    @property
    def cycle_count(self) -> int:
        """Total cycles recorded."""
        return self._cycle_count

    @property
    def is_mature(self) -> bool:
        """Whether the gate has reached its terminal (REJECT) state."""
        return self._current == GatePolicy.REJECT

File: bizra-omega/bizra-core/src/gate_policy.rs
//! Constitutional Gate Policy — unified enforcement for threshold violations.
//!
//! This module defines the canonical enforcement protocol for all constitutional
//! threshold checks (Ihsan, SNR, ADL, etc.) across the BIZRA ecosystem.
//!
//! ## The Problem
//! The same threshold check (`ihsan < 0.95`) previously triggered 5 different
//! behaviors across modules: silent drop, hard halt, zero reward, policy-dependent,
//! and APPROVED/REJECTED decisions. This module unifies the decision point.
//!
//! ## Usage
//! ```rust
//! use bizra_core::gate_policy::{apply_gate, env_gate_policy, GateAction, GatePolicy};
//!
//! let policy = env_gate_policy(); // Observe in dev, Reject in prod
//! let verdict = apply_gate(0.93, 0.95, policy);
//! match verdict.action {
//!     GateAction::Allow => { /* proceed */ }
//!     GateAction::AllowWithWarning => { /* log and proceed */ }
//!     GateAction::Rejected => { /* fail-closed */ }
//!     _ => { /* handle Flag/Throttle */ }
//! }
//! ```
//!
//! Standing on Giants: Al-Ghazali (Ihsan as obligation) · Deming (quality at source)

/// What happens when a constitutional threshold is violated.
///
/// Ordered from most permissive to most restrictive:
/// Observe → Flag → Throttle(n) → Reject
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum GatePolicy {
    /// Log warning but allow (development/observation mode)
    #[default]
    Observe,
    /// Attach warning flag to the event, still deliver
    Flag,
    /// Allow 1 in N events from the violating component
    Throttle(u32),
    /// Hard reject — fail-closed (production mode)
    Reject,
}

/// The action taken after policy evaluation.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GateAction {
    /// Score passed threshold — proceed normally
    Allow,
    /// Score failed but policy is Observe — proceed with logged warning
    AllowWithWarning,
    /// Score failed, policy is Flag — proceed with flag attached
    Flagged,
    /// Score failed, policy is Throttle — suppressed this time
    Throttled,
    /// Score failed, policy is Reject — hard stop
    Rejected,
}

/// The result of applying a gate policy to a score.
#[derive(Debug, Clone)]
pub struct GateVerdict {
    /// The score that was evaluated
    pub score: f64,
    /// The threshold it was measured against
    pub threshold: f64,
    /// Whether the score met the threshold
    pub passed: bool,
    /// The policy that was applied
    pub policy: GatePolicy,
    /// The resulting action
    pub action: GateAction,
}

/// Resolve the active gate policy from the `BIZRA_ENV` environment variable.
///
/// - `BIZRA_ENV=prod` or `BIZRA_ENV=production` → `GatePolicy::Reject`
/// - All other values or unset → `GatePolicy::Observe`
///
/// Individual modules MAY override this with explicit configuration.
pub fn env_gate_policy() -> GatePolicy {
    match std::env::var("BIZRA_ENV").as_deref() {
        Ok("prod") | Ok("production") => GatePolicy::Reject,
        _ => GatePolicy::Observe,
    }
}

/// Apply a gate policy to a score/threshold pair.
///
/// This is the canonical function all enforcement surfaces should call.
///
/// - If `score >= threshold`, returns `GateAction::Allow` regardless of policy.
/// - If `score < threshold`, the policy determines the action.
///
/// **Throttle counter state** is the caller's responsibility — this function
/// always returns `Throttled` for the `Throttle` policy on failure. The caller
/// decides whether this particular invocation is the 1-in-N that gets through.
pub fn apply_gate(score: f64, threshold: f64, policy: GatePolicy) -> GateVerdict {
    let passed = score >= threshold;
    let action = if passed {
        GateAction::Allow
    } else {
        match policy {
            GatePolicy::Observe => GateAction::AllowWithWarning,
            GatePolicy::Flag => GateAction::Flagged,
            GatePolicy::Throttle(_) => GateAction::Throttled,
            GatePolicy::Reject => GateAction::Rejected,
        }
    };
    GateVerdict {
        score,
        threshold,
        passed,
        policy,
        action,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::Mutex;

    // Env-var tests mutate global state — serialize them to avoid thread races.
    static ENV_LOCK: Mutex<()> = Mutex::new(());

    #[test]
    fn test_passing_score_always_allows() {
        let v = apply_gate(0.96, 0.95, GatePolicy::Reject);
        assert!(v.passed);
        assert_eq!(v.action, GateAction::Allow);
    }

    #[test]
    fn test_failing_observe_allows_with_warning() {
        let v = apply_gate(0.90, 0.95, GatePolicy::Observe);
        assert!(!v.passed);
        assert_eq!(v.action, GateAction::AllowWithWarning);
    }

    #[test]
    fn test_failing_flag_returns_flagged() {
        let v = apply_gate(0.90, 0.95, GatePolicy::Flag);
        assert!(!v.passed);
        assert_eq!(v.action, GateAction::Flagged);
    }

    #[test]
    fn test_failing_throttle_returns_throttled() {
        let v = apply_gate(0.90, 0.95, GatePolicy::Throttle(3));
        assert!(!v.passed);
        assert_eq!(v.action, GateAction::Throttled);
    }

    #[test]
    fn test_failing_reject_returns_rejected() {
        let v = apply_gate(0.90, 0.95, GatePolicy::Reject);
        assert!(!v.passed);
        assert_eq!(v.action, GateAction::Rejected);
    }

    #[test]
    fn test_exact_threshold_passes() {
        let v = apply_gate(0.95, 0.95, GatePolicy::Reject);
        assert!(v.passed);
        assert_eq!(v.action, GateAction::Allow);
    }

    #[test]
    fn test_env_gate_default_is_observe() {
        let _lock = ENV_LOCK.lock().unwrap();
        let prev = std::env::var("BIZRA_ENV").ok();
        std::env::remove_var("BIZRA_ENV");
        let p = env_gate_policy();
        // Restore
        match prev {
            Some(v) => std::env::set_var("BIZRA_ENV", v),
            None => std::env::remove_var("BIZRA_ENV"),
        }
        assert_eq!(p, GatePolicy::Observe);
    }

    #[test]
    fn test_env_gate_prod_is_reject() {
        let _lock = ENV_LOCK.lock().unwrap();
        let prev = std::env::var("BIZRA_ENV").ok();
        std::env::set_var("BIZRA_ENV", "prod");
        let p = env_gate_policy();
        // Restore
        match prev {
            Some(v) => std::env::set_var("BIZRA_ENV", v),
            None => std::env::remove_var("BIZRA_ENV"),
        }
        assert_eq!(p, GatePolicy::Reject);
    }

    #[test]
    fn test_env_gate_production_long_form() {
        let _lock = ENV_LOCK.lock().unwrap();
        let prev = std::env::var("BIZRA_ENV").ok();
        std::env::set_var("BIZRA_ENV", "production");
        let p = env_gate_policy();
        // Restore
        match prev {
            Some(v) => std::env::set_var("BIZRA_ENV", v),
            None => std::env::remove_var("BIZRA_ENV"),
        }
        assert_eq!(p, GatePolicy::Reject);
    }

    #[test]
    fn test_verdict_fields_populated() {
        let v = apply_gate(0.93, 0.95, GatePolicy::Reject);
        assert!((v.score - 0.93).abs() < f64::EPSILON);
        assert!((v.threshold - 0.95).abs() < f64::EPSILON);
        assert!(!v.passed);
        assert_eq!(v.policy, GatePolicy::Reject);
        assert_eq!(v.action, GateAction::Rejected);
    }
}
